### ISSUE TITLE
opt: hoist uncorrelated scalar subquery with equality

### ISSUE THREAD

**Issue by nvanbenschoten** [cockroachdb/cockroach#83392]
The following query plan shows a simple point lookup:
```sql
root@127.0.0.1:26257/movr> explain select * from promo_codes where code = 'test';

  • scan
    estimated row count: 1 (<0.01% of the table; stats collected 23 minutes ago)
    table: promo_codes@promo_codes_pkey
    spans: [/'test' - /'test']
```

When a subquery is used in an IN predicate, we can decorrelate the subquery and turn the predicate into a join. Arguably the join is redundant and we could eliminate it in this case because the tables on both sides are the same, but this is good enough.

```sql
root@127.0.0.1:26257/movr> explain select * from promo_codes where code IN (select code from promo_codes where code = 'test');

  • cross join (semi)
  │ estimated row count: 1
  │
  ├── • scan
  │     estimated row count: 1 (<0.01% of the table; stats collected 23 minutes ago)
  │     table: promo_codes@promo_codes_pkey
  │     spans: [/'test' - /'test']
  │
  └── • scan
        estimated row count: 1 (<0.01% of the table; stats collected 23 minutes ago)
        table: promo_codes@promo_codes_pkey
        spans: [/'test' - /'test']
```

However, when a subquery is used in an equality predicate, we can't decorrelate the subquery. Instead, we evaluate it, then perform a full table scan, then filter the result of that scan using the result of the subquery.

```sql
root@127.0.0.1:26257/movr> explain select * from promo_codes where code = (select code from promo_codes where code = 'test');

  • root
  │
  ├── • filter
  │   │ estimated row count: 333,664
  │   │ filter: code = @S1
  │   │
  │   └── • scan
  │         estimated row count: 1,000,991 (100% of the table; stats collected 26 minutes ago)
  │         table: promo_codes@promo_codes_pkey
  │         spans: FULL SCAN
  │
  └── • subquery
      │ id: @S1
      │ original sql: (SELECT code FROM promo_codes WHERE code = 'test')
      │ exec mode: one row
      │
      └── • scan
            estimated row count: 1 (<0.01% of the table; stats collected 26 minutes ago)
            table: promo_codes@promo_codes_pkey
            spans: [/'test' - /'test']
```

This is unfortunate and may look artificial, but while flying over the rocky mountains, @andreimatei and I found this by looking at real queries, like the following:
```sql
root@127.0.0.1:26257/movr> explain delete from promo_codes where code = (select code from promo_codes limit 1);

  • root
  │
  ├── • delete
  │   │ from: promo_codes
  │   │ auto commit
  │   │
  │   └── • filter
  │       │ estimated row count: 333,664
  │       │ filter: code = @S1
  │       │
  │       └── • scan
  │             estimated row count: 1,000,991 (100% of the table; stats collected 35 minutes ago)
  │             table: promo_codes@promo_codes_pkey
  │             spans: FULL SCAN
  │
  └── • subquery
      │ id: @S1
      │ original sql: (SELECT code FROM promo_codes LIMIT 1)
      │ exec mode: one row
      │
      └── • scan
            estimated row count: 1 (<0.01% of the table; stats collected 35 minutes ago)
            table: promo_codes@promo_codes_pkey
            spans: LIMITED SCAN
            limit: 1
```

Jira issue: CRDB-17044

**Comment by mgartner** [comment]
With your `=` example, the subquery is a scalar subquery. I don't think there's a way to decorrelate it like we do with `IN (<subquery>)`. It would certainly be nice to replan the main query based on the values fetched from the subquery.

However, in your `DELETE` example, it sounds like the best solution is for us to implement `DELETE FROM ... USING` so you don't have to use a subquery at all. https://github.com/cockroachdb/cockroach/issues/40963

**Comment by andreimatei** [comment]
Spell it out to me more why we can decorrelate the `IN` but not the `=`. Whatever reason there is, is it a good one?

**Comment by nvanbenschoten** [comment]
Does it have to do with the need to assert that the subquery does not return more than 1 row?

**Comment by mgartner** [comment]
> Spell it out to me more why we can decorrelate the IN but not the =. Whatever reason there is, is it a good one?

I was wrong about this. `=` is the same as `IN (...)` with a single item, so we should be able to decorrelate it.

> Does it have to do with the need to assert that the subquery does not return more than 1 row?

As long as we use a `Max1RowExpr` on the right side of the semi-join, decorrelation should be valid.

**Comment by andy-kimball** [comment]
Today, we only perform decorrelation for _correlated_ subqueries. Each subquery example here is not correlated, and so we execute it using a subquery rather than a left join. If we decide to change our execution strategy here, we'll need to be careful that we don't regress performance, as there may be cases where the subquery plan is better than the left join plan. @mgartner, isn't this a dup of https://github.com/cockroachdb/cockroach/issues/51820 and friends?

A bit of history: I deliberately didn't use left join for uncorrelated subqueries b/c I was afraid of regressing perf from the heuristic planner. I didn't have the same concern about correlated subqueries, b/c the heuristic planner didn't support those, so there was no possible backwards-compat perf problem.

**Comment by nvanbenschoten** [comment]
>  Today, we only perform decorrelation for correlated subqueries. Each subquery example here is not correlated, and so we execute it using a subquery rather than a left join. 

But we do decorrelate the subquery into a join in the `code IN (select ...)` case. How does that differ from the `code = (select ...)` case?

**Comment by andy-kimball** [comment]
I believe it's because `IN` is a "multi-row subquery" construct whereas `=` is a "single-row subquery" construct. The `IN` case can return multiple rows, and so we always map it to the `Any` operator (as are NOT IN, SOME, ANY, and ALL operators), which is always turned into a left join. By contrast, the `=` case will error if it returns more than one row, and so we treat it differently in the optimizer.

While you know that the two cases are the same in this case b/c `code` is a primary key, the `optbuilder`, which does the mapping from `IN` and `=` does not have that information, and so must choose how to build the operators based solely on syntax.

**Comment by andreimatei** [comment]
> By contrast, the = case will error if it returns more than one row, and so we treat it differently in the optimizer.

I think this is where the `Max1RowExpr` operator that @mgartner mentioned would come in, right?

FWIW, I think it's pretty important to optimize this form of query. I think we got to it from a TPCC query, but now I'm not sure. In any case, I've definitely seen people write this query.

**Comment by andy-kimball** [comment]
The optimizer would have no trouble turning `=` into a left-join query - that's not the problem. The main thing we'd want to verify is that changing the behavior here to address this important case doesn't regress other important cases. If we convince ourselves of that, "fixing" this is near trivial.

**Comment by mgartner** [comment]
I've thought about this a bit more, and I'm now fairly convinced that hoisting uncorrelated subqueries should _mostly_ lead to query plans that are no slower than equivalent plans with unhoisted, uncorrelated subqueries.

Hoisting a subquery transforms it into an expression with an apply-join. The code comment @andy-kimball quoted above is pointing out that subquery execution is faster than apply-join execution because the results of the subquery are cached and can be scanned multiple times. So, hoisting an uncorrelated subquery would only lead to a worse plan if the resulting apply-join executed the RHS more than once or the apply-join was not reduced to a regular join.

Apply-joins resulting from hoisting uncorrelated subqueries are usually uncorrelated as well, because the source subquery was. If the apply-join is uncorrelated, it is trivially reduced to a regular join. This is the most common outcome of hoisting an uncorrelated subquery, and the performance would be improved because the subquery has been removed and no apply-join remains, likely enabling further optimizations (like using a lookup join or merge join with the `select * from promo_codes where code = (select code from promo_codes where code = 'test')` above, assuming an index on `promo_coes.code` exists).

The `HoistValuesSubquery` rule, however, would create correlated apply-joins when operating on an uncorrelated subquery. The rule creates a values expression on the RHS of an inner-apply-join that references a column in the LHS: https://github.com/cockroachdb/cockroach/blob/2675c7c0481bb33eb9f43f1c2b52c152fca3ae2d/pkg/sql/opt/norm/decorrelate_funcs.go#L264

Here's an example showing the inner-apply-join that remains after `HoistValuesSubquery` fires:

```
norm
SELECT v FROM (VALUES (1), ((SELECT c FROM c WHERE c=2))) f(v)
----
project
 ├── columns: v:5
 ├── cardinality: [2 - 2]
 └── inner-join-apply
      ├── columns: c:1 column1:5
      ├── cardinality: [2 - 2]
      ├── fd: ()-->(1)
      ├── left-join (cross)
      │    ├── columns: c:1
      │    ├── cardinality: [1 - 1]
      │    ├── multiplicity: left-rows(exactly-one), right-rows(exactly-one)
      │    ├── key: ()
      │    ├── fd: ()-->(1)
      │    ├── values
      │    │    ├── cardinality: [1 - 1]
      │    │    ├── key: ()
      │    │    └── ()
      │    ├── max1-row
      │    │    ├── columns: c:1!null
      │    │    ├── error: "more than one row returned by a subquery used as an expression"
      │    │    ├── cardinality: [0 - 1]
      │    │    ├── key: ()
      │    │    ├── fd: ()-->(1)
      │    │    └── select
      │    │         ├── columns: c:1!null
      │    │         ├── fd: ()-->(1)
      │    │         ├── scan c
      │    │         │    └── columns: c:1
      │    │         └── filters
      │    │              └── c:1 = 2 [outer=(1), constraints=(/1: [/2 - /2]; tight), fd=()-->(1)]
      │    └── filters (true)
      ├── values
      │    ├── columns: column1:5
      │    ├── outer: (1)
      │    ├── cardinality: [2 - 2]
      │    ├── (1,)
      │    └── (c:1,)
      └── filters (true)
```

This is a rare case, though, and I think it's unlikely to regress performance significantly. The resulting apply-join should execute the RHS only once, because the cardinality of the LHS is 1: the LHS produces exactly one row from a single-row values expression left-cross-joined with the subquery input, which produces 0 or 1 rows. Since the RHS of the apply-join is executed only once, the runtime should be similar to executing it as a subquery - there is no benefit of a cache from a single invocation. However, plans with subqueries can be distributed, while plans with apply-joins cannot.

In summary, if we hoist uncorrelated subqueries, I think we'd get better query plans in the majority of cases. In the rare case that `HoistValuesSubquery` triggers and creates an apply-join, we'd get a logical query plan with the same complexity as before, but we'd lose the ability to physically distribute the execution of the query.

To be safe, we can add a session setting that customers can use to revert to the previous behavior. If we think the `HoistValuesSubquery` case is more common than I'm assuming, then we can explore ways to disable that rule for uncorrelated subqueries.

**Comment by mgartner** [comment]
While working on #98375, I did notice that for `ANY` subqueries, hoisting can lead to a significantly less efficient query plan.

**Comment by mgartner** [comment]
I'm going to try to enabling hoisting of a subset of these types subqueries, behind a session setting so we can backport it. to 22.2 and 23.1.
